cmd/add.go

```diff
@@ -1,54 +1,160 @@
+	"github.com/rs/zerolog/log"
 	"github.com/spf13/cobra"
 )
 
 func addAdd(cmd *cobra.Command) {
 	addCmd := &cobra.Command{
-		Use:   "add",
+		Use:   "add [url]",
 		Short: "Add new resource",
 		Args:  cobra.MinimumNArgs(1),
 		RunE:  runAdd,
 	}
+
+	addCmd.Flags().String("cache", "", "Artifactory cache URL")
 	addCmd.Flags().String("algo", internal.RecommendedAlgo, "Integrity algorithm")
-	addCmd.Flags().String("filename", "", "Target file name to use when downloading the resource")
+	addCmd.Flags().String("filename", "", "Target file name")
 	addCmd.Flags().StringArray("tag", []string{}, "Resource tags")
+
 	cmd.AddCommand(addCmd)
 }
 
 func runAdd(cmd *cobra.Command, args []string) error {
+	// Get flags
 	lockFile, err := cmd.Flags().GetString("lock-file")
 	if err != nil {
 		return err
 	}
-	lock, err := internal.NewLock(lockFile, true)
-	if err != nil {
-		return err
-	}
+
 	algo, err := cmd.Flags().GetString("algo")
 	if err != nil {
 		return err
 	}
+
 	tags, err := cmd.Flags().GetStringArray("tag")
 	if err != nil {
 		return err
 	}
+
 	filename, err := cmd.Flags().GetString("filename")
 	if err != nil {
 		return err
 	}
-	err = lock.AddResource(args, algo, tags, filename)
+
+	cache, err := cmd.Flags().GetString("cache")
 	if err != nil {
 		return err
 	}
-	err = lock.Save()
+
+	// Check for GRABIT_ARTIFACTORY_TOKEN if cache is specified
+	if cache != "" {
+		token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+		if token == "" {
+			return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN must be set when using cache")
+		}
+	}
+
+	// Create or open lock file
+	lock, err := internal.NewLock(lockFile, true)
 	if err != nil {
 		return err
 	}
+
+	// Download file first
+	ctx := context.Background()
+	tempFile, err := internal.GetUrltoTempFile(args[0], ctx)
+	if err != nil {
+		return fmt.Errorf("failed to download resource: %w", err)
+	}
+	defer os.Remove(tempFile)
+
+	// Calculate hash for integrity and cache path
+	hash, err := internal.GetFileHash(tempFile)
+	if err != nil {
+		return fmt.Errorf("failed to calculate file hash: %w", err)
+	}
+
+	// Handle cache if specified
+	var cachePath string
+	if cache != "" {
+		// Ensure cache URL ends with a single /
+		cache = strings.TrimSuffix(cache, "/") + "/"
+		cachePath = cache + hash
+
+		log.Debug().
+			Str("path", cachePath).
+			Msg("Uploading to Artifactory cache")
+
+		if err := uploadToArtifactory(tempFile, cachePath); err != nil {
+			log.Debug().
+				Err(err).
+				Str("path", cachePath).
+				Msg("Failed to upload to cache")
+			// Continue without cache - this is not a fatal error
+		} else {
+			log.Debug().
+				Str("path", cachePath).
+				Msg("Successfully uploaded to cache")
+		}
+	}
+
+	// Add resource to lock file
+	if err := lock.AddResourceWithCache(args, algo, tags, filename, cachePath); err != nil {
+		return fmt.Errorf("failed to add resource to lock file: %w", err)
+	}
+
+	return lock.Save()
+}
+
+func uploadToArtifactory(filePath, cacheUrl string) error {
```

Comment:
This function is only called from one location, and that location already had done the `os.Getenv()` to get the token env variable.  It might be better to pass in the token as a param, rather than `os.Getenv()`'ing again.